model/src/main/kotlin/io/spine/validation/SetOnceOption.kt

```diff
@@ -0,0 +1,127 @@
+import io.spine.protodata.ast.name
+import io.spine.protodata.ast.qualifiedName
+import io.spine.protodata.ast.ref
+import io.spine.protodata.check
+import io.spine.protodata.plugin.Policy
+import io.spine.protodata.plugin.View
+import io.spine.server.entity.alter
+import io.spine.server.event.NoReaction
+import io.spine.server.event.React
+import io.spine.server.event.asA
+import io.spine.server.tuple.EitherOf2
+import io.spine.validation.event.SetOnceFieldDiscovered
+import io.spine.validation.event.setOnceFieldDiscovered
+
+/**
+ * Controls whether a field should be validated with the `(set_once)` option.
+ *
+ * Whenever a field marked with `(set_once)` option is discovered, emits
+ * [SetOnceFieldDiscovered] event if the following conditions are met:
+ *
+ * 1. The field type is supported by the option.
+ * 2. The option value is `true`.
+ *
+ * If (1) is violated, the policy reports a compilation error.
+ *
+ * Violation of (2) means that the `(set_once)` option is applied correctly,
+ * but disabled. In this case, the policy emits [NoReaction] because we actually
+ * have a non-set-once field, marked with `(set_once)`.
+ */
+internal class SetOncePolicy : Policy<FieldOptionDiscovered>() {
+
+    @React
+    override fun whenever(
+        @External @Where(field = OPTION_NAME, equals = SET_ONCE)
+        event: FieldOptionDiscovered
+    ): EitherOf2<SetOnceFieldDiscovered, NoReaction> {
+        val field = event.subject
+        val file = event.file
+        checkFieldType(field, file)
+
+        if (!event.option.boolValue) {
+            return ignore()
+        }
+
+        val message = resolveErrorMessage<IfSetAgainOption>(field)
+        return setOnceFieldDiscovered {
+            id = field.ref
+            errorMessage = message
+            subject = field
+        }.asA()
+    }
+}
+
+/**
+ * Reports a compilation error when the `(if_set_again)` option is applied
+ * without `(set_once)`.
+ */
+internal class IfSetAgainPolicy : CompanionPolicy(
+    primary = OptionsProto.setOnce,
+    companion = OptionsProto.ifSetAgain,
+) {
+    @React
+    override fun whenever(@External event: FieldOptionDiscovered) = checkWithPrimary(event)
+}
+
+/**
+ * A view of a field that is marked with `(set_once) = true` option.
+ */
+internal class SetOnceFieldView : View<FieldRef, SetOnceField, SetOnceField.Builder>() {
+
+    @Subscribe
+    fun on(e: SetOnceFieldDiscovered) = alter {
+        errorMessage = e.errorMessage
+        subject = e.subject
+    }
+}
+
+private fun checkFieldType(field: Field, file: File) =
+    Compilation.check(field.type.isSupported(), file, field.span) {
+        "The field type `${field.type.name}` of the `${field.qualifiedName}` field" +
+                " is not supported by the `($SET_ONCE)` option. Supported field types: messages," +
+                " enums, strings, bytes, bools and all numeric fields."
+    }
+
+private fun FieldType.isSupported(): Boolean = !isList && !isMap
```

Comment:
Please document this extension. Supported for what? Supported by whom?

Reply:
I've added the comment, but GitHub doesn't mark it as `outdated` for some reason.